File: DsaWithCpp/Heap/KSortedArray.cpp
#include <iostream>
#include "PriorityQueue.cpp"
using namespace std;

static void kSort(int arr[], int k, int n)
{
    PriorityQueue pq;
    for (int i = 0; i <= k; i++)
        pq.push(arr[i]);
    int index = 0;
    for (int i = k + 1; i < n; i++)
    {
        arr[index++] = pq.pop();
        pq.push(arr[i]);
    }
    while (!pq.isEmpty())
        arr[index++] = pq.pop();
}

// int main()
// {
//     int arr[] = {9, 8, 7, 19, 18};
//     kSort(arr, 3, 6);
//     for (int i = 0; i < 5; i++)
//         cout << arr[i] << " ";
//     cout << endl;
// }

File: Python/heap/k_sorted_array.py
import priority_queue as pq


def k_sort(arr: list, k):
    for i in range(k + 1):
        pq.add(arr[i])
    idx = 0
    for i in range(k + 1, len(arr)):
        arr[idx] = pq.poll()
        idx += 1
        pq.add(arr[i])
    while not pq.is_empty():
        arr[idx] = pq.poll()
        idx += 1


def main():
    arr = [9, 8, 7, 19, 18]
    k_sort(arr, 3)
    print(arr)


if __name__ == "__main__":
    main()
